libs/datamodel/core/src/transform/ast_to_dml/db/walkers/relation.rs

```diff
@@ -1,15 +1,136 @@
+use super::{ModelWalker, RelationFieldWalker, ScalarFieldWalker};
+use crate::{
+    ast,
+    transform::ast_to_dml::db::{relations::*, ParserDatabase},
+};
 use datamodel_connector::ConnectorCapability;
 use dml::relation_info::ReferentialAction;
 
-use crate::{
-    ast::{self, FieldArity},
-    transform::ast_to_dml::db::{
-        relations::{Relation, RelationType},
-        ParserDatabase,
-    },
-};
+#[derive(Copy, Clone)]
+pub(crate) struct RelationWalker<'ast, 'db> {
+    pub(super) relation_id: usize,
+    pub(super) db: &'db ParserDatabase<'ast>,
+}
 
-use super::{ModelWalker, RelationFieldWalker, ScalarFieldWalker};
+impl<'ast, 'db> RelationWalker<'ast, 'db> {
+    pub(crate) fn refine(&self) -> RefinedRelationWalker<'ast, 'db> {
+        let relation = &self.db.relations.relations_storage[self.relation_id];
+
+        if relation.is_many_to_many() {
+            RefinedRelationWalker::ImplicitManyToMany(ImplicitManyToManyRelationWalker {
+                db: self.db,
+                relation_id: self.relation_id,
+            })
+        } else {
+            RefinedRelationWalker::Inline(InlineRelationWalker {
+                relation_id: self.relation_id,
+                db: self.db,
+            })
+        }
+    }
+}
+
+pub(crate) enum RefinedRelationWalker<'ast, 'db> {
+    Inline(InlineRelationWalker<'ast, 'db>),
+    ImplicitManyToMany(ImplicitManyToManyRelationWalker<'ast, 'db>),
+}
+
+#[derive(Copy, Clone)]
+pub(crate) struct InlineRelationWalker<'ast, 'db> {
+    relation_id: usize,
+    db: &'db ParserDatabase<'ast>,
+}
+
+impl<'ast, 'db> InlineRelationWalker<'ast, 'db> {
+    fn get(&self) -> &'db Relation<'ast> {
```

Comment:
Get what? What about `fn attributes` or `fn relation`?

Reply:
Yeah, we can iterate on this. Not set to anything so if we standardize, let's make it the same everywhere. The question though is, what should it be for things backed by ast. The attributes, or the ast data?

Reply:
Depends — `RelationWalker` doesn't refer to anything from the AST, but the rest should use the AST, since everything else uses AST ids as identifiers.